### ISSUE TITLE
Commands out of sync; you can't run this command now

### ISSUE THREAD

**Issue by ziutek** [Philio/GoMySQL#10]
I get error 2014: "Commands out of sync; you can't run this command now" when execute query like:
    SELECT String FROM Table
There is 2270 rows in Table. String is varchar(50). When I set MySQL.Logging to true and add this line to  func (mysql *MySQL) getResult() (err os.Error):
    fmt.Printf("MCH mysql: seq=%d, hdr: seq=%d len=%d\n", mysql.sequence, hdr.sequence, hdr.length)
I get:
    [cut]
    MCH mysql: seq=98, hdr: seq=98 len=7
    2010/08/18 20:45:22 [98] Received row data packet from server
    MCH mysql: seq=99, hdr: seq=99 len=10
    2010/08/18 20:45:22 [99] Received row data packet from server
    MCH mysql: seq=100, hdr: seq=100 len=9
    2010/08/18 20:45:22 [100] Received row data packet from server
    MCH mysql: seq=101, hdr: seq=101 len=7
    2010/08/18 20:45:22 [101] Received row data packet from server
    MCH mysql: seq=102, hdr: seq=102 len=1
    2010/08/18 20:45:22 [102] Received ok packet from server
    2010/08/18 20:45:22 Current result set saved
    MCH mysql: seq=103, hdr: seq=197 len=4544321
    2010/08/18 20:45:22 DB error 2014: Commands out of sync; you can't run this command now
When I change order in result, modifying query like this:
    SELECT String FROM Table ORDER BY String
I get same error but less logs:
    2010/08/18 20:47:00 Connect called
    2010/08/18 20:47:00 Connected using TCP/IP
    2010/08/18 20:47:00 [0] Received init packet from server
    2010/08/18 20:47:00 [1] Sent auth packet to server
    MCH mysql: seq=2, hdr: seq=2 len=7
    2010/08/18 20:47:00 [2] Received ok packet from server
    2010/08/18 20:47:00 Current result set saved
    2010/08/18 20:47:00 Query called with SQL: set names utf8
    2010/08/18 20:47:00 [0] Sent query command to server
    MCH mysql: seq=1, hdr: seq=1 len=7
    2010/08/18 20:47:00 [1] Received ok packet from server
    2010/08/18 20:47:00 Current result set saved
    2010/08/18 20:47:00 web.go serving 10.128.101.7:8080
    2010/08/18 20:47:03 GET /
    2010/08/18 20:47:03 Query called with SQL: select imie from Osoby order by imie
    2010/08/18 20:47:03 [0] Sent query command to server
    MCH mysql: seq=1, hdr: seq=1 len=1
    2010/08/18 20:47:03 [1] Received result set packet from server
    MCH mysql: seq=2, hdr: seq=2 len=45
    2010/08/18 20:47:03 [2] Received field packet from server
    MCH mysql: seq=3, hdr: seq=3 len=5
    2010/08/18 20:47:03 [3] Received eof packet from server
    2010/08/18 20:47:03 End of field packets
    MCH mysql: seq=4, hdr: seq=4 len=1
    2010/08/18 20:47:03 [4] Received ok packet from server
    2010/08/18 20:47:03 Current result set saved
    MCH mysql: seq=5, hdr: seq=0 len=393216
    2010/08/18 20:47:03 DB error 2014: Commands out of sync; you can't run this command now
When I add primary key to SELECT, like this:
    SELECT Id, String FROM Table
I get no errors.

There is 21 fields in My real table. I get this error when I select any varchar field or fields, without primary key field. There is no errors when I select date, datetime or enum fields.

**Comment by Philio** [comment]
This is strange, I've never had any issue like this from a simple select. Can you provide the table structures please?

**Comment by ziutek** [comment]
There is table structure with modified field names (security reasons):

```
CREATE TABLE `SomeTable` (
  `id` int(2) unsigned NOT NULL auto_increment,
  `s1` varchar(80) NOT NULL default '',
  `s2` varchar(50) NOT NULL default '',
  `s3` varchar(24) NOT NULL,
  `s4` varchar(20) NOT NULL default '',
  `s5` varchar(9) default NULL,
  `s6` varchar(5) default NULL,
  `e1` enum('T','N') NOT NULL default 'T',
  `s7` varchar(32) default NULL,
  `d1` date NOT NULL default '0000-00-00',
  `s8` varchar(80) default NULL,
  `s9` varchar(80) default NULL,
  `s10` varchar(16) NOT NULL default '',
  `d2` date default NULL,
  `d3` date default NULL,
  `d4` date default NULL,
  `d5` date default NULL,
  `d6` date default NULL,
  `e2` enum('T','N','K') NOT NULL default 'T',
  `d7` datetime NOT NULL default '0000-00-00 00:00:00',
  `s11` varchar(24) NOT NULL default '',
  PRIMARY KEY  (`id`)
) ENGINE=MyISAM AUTO_INCREMENT=2339 DEFAULT CHARSET=latin2
```

My MySQL version is: 5.0.51a-24+lenny4-log

I get no errors when select is limited to <300 entries.

**Comment by Philio** [comment]
Thanks, will check this out when I get a chance!

**Comment by ziutek** [comment]
I tried thoj's Go-MySQL-Client-Library. His quickly written code contains a lot of hacks, and poor error checking but after fix to Reader.Read usage it works without this issue (without my fix it doesn't work for long results).

**Comment by Philio** [comment]
This should be resolved in the latest release.

**Comment by thomaslee** [comment]
I spoke to ziutek briefly about this issue on the go-nuts mailing list. He had this to say about the issue:

"I realized during writing my package that this bug appears when text result row has NULL value in its first column. GoMySQL treat such Row Data Packet as OK packet and goes into the forest - Polish idiom ;). But this bug can be avoided by using prepared statements."

So this bug should manifest itself when the first column in a result set is NULL. Haven't attempted to reproduce this yet.

**Comment by thomaslee** [comment]
Will open a separate issue if this can be reproduced.

**Comment by Philio** [comment]
I've reopened :)

**Comment by ziutek** [comment]
I made a mistake in the sentence: "bug appears when text result row has NULL value in its first column". I meant an empty string, not a NULL value. NULL in first column gives 251 value in first byte after packet header, and it isn't problem. Empty string gives 0 value in first byte after header and it is treated by GoMySQL as mark of OK packet.

**Comment by Philio** [comment]
Ah right, yer that makes sense, as there is no check in place to make sure that the OK packet is the first packet in the set. Of course OK should only occur if it is the first packet!

**Comment by Philio** [comment]
Fixed, 0.2.5
